http: don't emit 'data' after 'error'

ronag opened the pull request:
Don't emit pending 'data' after 'error'.

##### Checklist

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

jasnell commented:
Defensively marking as semver-major due to the change in event timing. I'd prefer to view this as a bug fix tho. @mcollina what do you think?

ronag commented:
@mcollina: I'm just trying to improve consistency and predictability where it makes sense and risk is reasonable.

The implicit assumption for streams is (as we discussed elsewhere) that we only emit `close` after `error`.

Here is a potential case where this can be a problem:

```js

req
 .on('error', err => {
   fs.close(file)
 })
 .on('data', buf => {
   // uh oh... we might already be closed...
   fs.write(file, buf)
 })
```

ronag commented:
> I actually don’t agree with this order. Arguably knowing that there was an error and marking the socket is more important than emitting whatever data was leftover.

The alternative is that we just dump any pending data. Makes sense to me. Should I change it?

mcollina commented:
I would prefer if we dumped any pending data. It actually make more sense.

ronag commented:
Updated

ronag commented:
@apapirovski approve?

ronag commented:
@benjamingr this has stalled?

Trott commented:
> @benjamingr @Trott this has stalled? ping node/http?

@ronag Should the commit message be updated? This is more about not reading data on the error condition than ordering of events?

@jasnell Is this still a breaking change in your opinion?

@apapirovski Are you still opposed on the grounds that this has never been reported as a significant real-world problem? Or with the change as it is now, do you think this is something we should do? Similar question for @mcollina 

/ping @nodes/http for review/comments on above questions

ronag commented:
@Trott: Rebased and fixed message.

I **think** @mcollina and @apapirovski's objections were resolved with the current fix, i.e. don't emit data after error at all.

Trott commented:
/ping @nodejs/http for review/comments

Trott commented:
CITGM: https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/1962/

ronag commented:
@Trott: this looks ready

mcollina commented:
@apapirovski @lpinca what do you think about this change?

Trott commented:
CITGM: https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/2001/

Trott commented:
@apapirovski Would love to get your 👍 , 👎 , or ¯\\_(ツ)_/¯ on this

Trott commented:
CITGM results seem OK.

Trott commented:
> I actually don’t agree with this order. Arguably knowing that there was an error and marking the socket is more important than emitting whatever data was leftover.
> 
> I’m open to the alternative but it’s unclear what the benefit is — given there’s been no feedback on this over the years.

Given that this PR no longer emits data after error, I think @apapirovski's objection has been addressed. Multiple pings have not brought him back into this conversation, so I'm going to clear the objection based on https://github.com/nodejs/node/blob/a0c6cf8eb1bbe4b964b8267341e894a8fafeab51/COLLABORATOR_GUIDE.md#consensus-seeking.

Let's leave this open another 48 hours or so to see if @apapirovski or anyone else has anything more to say. (I'll ping @apapirovski through non-GitHub channels just in case the problem here is being overwhelmed by GitHub notifications or whatever.) But as of now, this has sufficient approvals, a good CTIGM run, and a good CI run.

Trott commented:
Landed in e573c39
